usr/src/man/man1/alias.1
.\"
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\"
.\" The Institute of Electrical and Electronics Engineers and The Open
.\" Group, have given us permission to reprint portions of their
.\" documentation.
.\"
.\" In the following statement, the phrase ``this text'' refers to portions
.\" of the system documentation.
.\"
.\" Portions of this text are reprinted and reproduced in electronic form
.\" in the SunOS Reference Manual, from IEEE Std 1003.1, 2004 Edition,
.\" Standard for Information Technology -- Portable Operating System
.\" Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group.  In the event of any discrepancy
.\" between these versions and the original IEEE and The Open Group
.\" Standard, the original IEEE and The Open Group Standard is the referee
.\" document.  The original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"
.\" This notice shall appear on any product containing this material.
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright 1989 AT&T
.\" Portions Copyright (c) 1992, X/Open Company Limited.  All Rights Reserved.
.\" Portions Copyright (c) 1982-2007 AT&T Knowledge Ventures
.\" Copyright (c) 2007, Sun Microsystems, Inc. All Rights Reserved.
.\"
.TH ALIAS 1 "Apr 8, 2008"
.SH NAME
alias, unalias \- create or remove a pseudonym or shorthand for a command or
series of commands
.SH SYNOPSIS
.LP
.nf
\fB/usr/bin/alias\fR [\fIalias-name\fR[= \fIstring\fR...]]
.fi

.LP
.nf
\fB/usr/bin/unalias\fR \fIalias-name\fR...
.fi

.LP
.nf
\fB/usr/bin/unalias\fR \fB-a\fR
.fi

.SS "\fBcsh\fR"
.LP
.nf
\fBalias\fR [\fIname\fR [\fIdef\fR]]
.fi

.LP
.nf
\fBunalias\fR \fIpattern\fR
.fi

.SS "\fBksh\fR"
.LP
.nf
\fBalias\fR [\fB-tx\fR] [\fIname\fR[= \fIvalue\fR]...]
.fi

.LP
.nf
\fBunalias\fR \fIname\fR...
.fi

.LP
.nf
\fBunalias\fR [\fB-a\fR]
.fi

.SS "\fBksh93\fR"
.LP
.nf
\fBalias\fR [\fB-ptx\fR] [\fIname\fR[= \fIvalue\fR]...]
.fi

.LP
.nf
\fBunalias\fR [\fB-a\fR] [\fIname\fR...]
.fi

.SH DESCRIPTION
.sp
.LP
The \fBalias\fR and \fBunalias\fR utilities create or remove a pseudonym or
shorthand term for a command or series of commands, with different
functionality in the C-shell and Korn shell environments.
.SS "\fB/usr/bin/alias\fR"
.sp
.LP
The \fBalias\fR utility creates or redefines alias definitions or writes the
values of existing alias definitions to standard output. An alias definition
provides a string value that replaces a command name when it is encountered.
.sp
.LP
An alias definition affects the current shell execution environment and the
execution environments of the subshells of the current shell. When used as
specified by this document, the alias definition does not affect the parent
process of the current shell nor any utility environment invoked by the shell.
.SS "\fB/usr/bin/unalias\fR"
.sp
.LP
The \fBunalias\fR utility removes the definition for each alias name specified.
The aliases are removed from the current shell execution environment. The
\fB-a\fR option removes all alias definitions from the current execution
environment.
.SS "\fBcsh\fR"
.sp
.LP
\fBalias\fR assigns \fIdef\fR to the alias \fIname\fR. The assigned \fIdef\fR
is a list of words that can contain escaped history-substitution metasyntax.
\fIname\fR is not allowed to be \fBalias\fR or \fBunalias\fR. If \fIdef\fR is
omitted, the alias \fIname\fR is displayed along with its current definition.
If both \fIname\fR and \fIdef\fR are omitted, all aliases are displayed.
.sp
.LP
Because of implementation restrictions, an alias definition must have been
entered on a previous command line before it can be used.
.sp
.LP
\fBunalias\fR discards aliases that match (filename substitution)
\fIpattern\fR. All aliases can be removed by `\fBunalias *\fR'.
.SS "\fBksh\fR"
.sp
.LP
\fBalias\fR with no arguments prints the list of aliases in the form
\fIname=value\fR on standard output. An \fBalias\fR is defined for each name
whose \fIvalue\fR is specified. A trailing space in \fIvalue\fR causes the next
word to be checked for alias substitution. The \fB-t\fR flag is used to set and
list tracked aliases. The value of a tracked alias is the full pathname
corresponding to the specified \fIname\fR. The value becomes undefined when the
value of \fBPATH\fR is reset but the aliases remained tracked. Without the
\fB-t\fR flag, for each \fIname\fR in the argument list for which no
\fIvalue\fR is specified, the name and value of the alias is printed. The
\fB-x\fR flag is used to set or print \fIexported alias\fRes. An exported alias
is defined for scripts invoked by \fIname\fR. The exit status is non-zero if a
\fIname\fR is specified, but no value, and no alias has been defined for the
\fIname\fR.
.sp
.LP
The \fBalias\fRs specified by the list of \fIname\fRs can be removed from the
\fBalias\fR list with \fBunalias\fR.
.SS "\fBksh93\fR"
.sp
.LP
\fBalias\fR creates or redefines alias definitions or writes the existing alias
definitions to standard output.
.sp
.LP
An alias definition provides a string value that replaces a command name when
the command is read. Alias names can contain any printable character that is
not special to the shell. If an alias value ends in a SPACE or TAB, the word
following the command name the alias replaces is also checked to see whether it
is an alias.
.sp
.LP
If no names are specified, the names and values of all aliases are written to
standard output. Otherwise, for each name that is specified, and
\fB=\fR\fIvalue\fR is not specified, the current value of the alias
corresponding to name is written to standard output. If \fB=\fR\fIvalue\fR is
specified, the alias name is created or redefined.
.sp
.LP
\fBalias\fR is built-in to the shell as a declaration command so that field
splitting and pathname expansion are not performed on the arguments. Tilde
expansion occurs on \fIvalue\fR. An alias definition only affects scripts read
by the current shell environment. It does not affect scripts run by this shell.
.sp
.LP
\fBunalias\fR removes the definition of each named alias from the current shell
execution environment, or all aliases if \fB-a\fR is specified. It does not
affect any commands that have already been read and subsequently executed.
.SH OPTIONS
.sp
.LP
The following option is supported by \fBunalias\fR:
.sp
.ne 2
.na
\fB\fB-a\fR\fR
.ad
.RS 6n
Removes all alias definitions from the current shell execution environment.
.RE

.SS "\fBksh\fR"
.sp
.LP
The following option is supported by \fBalias\fR:
.sp
.ne 2
.na
\fB\fB-t\fR\fR
.ad
.RS 6n
Sets and lists tracked aliases.
.RE

.SS "\fBksh93\fR"
.sp
.LP
The following options are supported by \fBalias\fR:
.sp
.ne 2
.na
\fB\fB-p\fR\fR
.ad
.RS 6n
Causes the output to be in the form of \fBalias\fR commands that can be used as
input to the shell to recreate the current aliases.
.RE

.sp
.ne 2
.na
\fB\fB-t\fR\fR
.ad
.RS 6n
Specifies tracked aliases.
.sp
Tracked aliases connect a command name to the command's pathname, and are reset
when the \fBPATH\fR variable is unset. The tracked aliases feature is now
obsolete.
.RE

.sp
.ne 2
.na
\fB\fB-x\fR\fR
.ad
.RS 6n
Ignored, this option is obsolete.
.RE

.sp
.LP
The following option is supported by \fBunalias\fR:
.sp
.ne 2
.na
\fB\fB-a\fR\fR
.ad
.RS 6n
Causes all alias definitions to be removed. \fIname\fR operands are optional
and ignored if specified.
.RE

.SH OPERANDS
.sp
.LP
The following operands are supported:
.SS "\fBalias\fR"
.sp
.ne 2
.na
\fB\fIalias-name\fR\fR
.ad
.RS 14n
Write the alias definition to standard output.
.RE

.SS "\fBunalias\fR"
.sp
.ne 2
.na
\fB\fIalias-name\fR\fR
.ad
.RS 21n
The name of an alias to be removed.
.RE

.sp
.ne 2
.na
\fB\fIalias-name\fR\fB=\fR\fIstring\fR\fR
.ad
.RS 21n
Assign the value of \fIstring\fR to the alias \fIalias-name\fR.
.RE

.sp
.LP
If no operands are specified, all alias definitions are written to standard
output.
.SH OUTPUT
.sp
.LP
The format for displaying aliases (when no operands or only \fIname\fR operands
are specified) is:
.sp
.in +2
.nf
"%s=%s\en" \fIname\fR, \fIvalue\fR
.fi
.in -2
.sp

.sp
.LP
The \fIvalue\fR string is written with appropriate quoting so that it is
suitable for reinput to the shell.
.SH EXAMPLES
.LP
\fBExample 1 \fRModifying a Command's Output
.sp
.LP
This example specifies that the output of the \fBls\fR utility is columnated
and more annotated:

.sp
.in +2
.nf
example% \fBalias ls="ls \(miCF"\fR
.fi
.in -2
.sp

.LP
\fBExample 2 \fRRepeating Previous Entries in the Command History File
.sp
.LP
This example creates a simple "redo" command to repeat previous entries in the
command history file:

.sp
.in +2
.nf
example% \fBalias r='fc \(mis'\fR
.fi
.in -2
.sp

.LP
\fBExample 3 \fRSpecifying a Command's Output Options
.sp
.LP
This example provides that the \fBdu\fR utility summarize disk output in units
of 1024 bytes:

.sp
.in +2
.nf
example% \fBalias du=du \(mik\fR
.fi
.in -2
.sp

.LP
\fBExample 4 \fRDealing with an Argument That is an Alias Name
.sp
.LP
This example sets up the \fBnohup\fR utility so that it can deal with an
argument that is an alias name:

.sp
.in +2
.nf
example% \fBalias nohup="nohup "\fR
.fi
.in -2
.sp

.SH ENVIRONMENT VARIABLES
.sp
.LP
See \fBenviron\fR(7) for descriptions of the following environment variables
that affect the execution of \fBalias\fR and \fBunalias\fR: \fBLANG\fR,
\fBLC_ALL\fR, \fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR\fR
.ad
.RS 5n
Successful completion.
.RE

.SS "\fBalias\fR"
.sp
.ne 2
.na
\fB\fB>0\fR\fR
.ad
.RS 6n
One of the \fIalias-name\fR operands specified did not have an alias
definition, or an error occurred.
.RE

.SS "\fBunalias\fR"
.sp
.ne 2
.na
\fB\fB>0\fR\fR
.ad
.RS 6n
One of the \fIalias-name\fR operands specified did not represent a valid alias
definition, or an error occurred.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(7) for descriptions of the following attributes:
.SS "\fBcsh, ksh\fR"
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
Interface Stability	Committed
_
Standard	See \fBstandards\fR(7).
.TE

.SS "\fBksh93\fR"
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
Interface Stability	Uncommitted
.TE

.SH SEE ALSO
.sp
.LP
.BR csh (1),
.BR ksh (1),
.BR ksh93 (1),
.BR shell_builtins (1),
.BR attributes (7),
.BR environ (7),
.BR standards (7)


usr/src/cmd/ksh/builtins/alias.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */

/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*
 * alias.c is a C version of the alias.sh wrapper (which links ksh
 * builtins to commands in /usr/bin/, e.g. calling this wrapper as
 * /usr/bin/alias will call the ksh "alias" builtin, running it as
 * /usr/bin/cut will call the ksh "cut" builtin etc.
 */

#include <shell.h>
#include <nval.h>
#include <cmdext.h>
#include <stdio.h>

typedef struct {
	const char *name;
	int (* func)(int, char **, void *);
} bfastpathrec;

/*
 * We've disabled the "fastpath" codepath for some commands below
 * because it causes a paradoxon for large input files (as used by
 * ON PerfPIT for testing). For /usr/bin/rev (where the issue was
 * first discovered) it looks like this:
 * - for small files like /etc/profile the fastpath is faster in a loop
 *   with 1000 iterations (8 seconds with fastpath, 14 seconds without
 *   fastpath)
 * - for large files (/usr/pub/UTF-8 replicated until the test file
 *   reaches 24884706 bytes) the benchmark reverses: The fastpath now
 *   needs 40 seconds and without fastpath it needs 30 seconds (for 100
 *   iterations).
 */
#if 0
#define	ENABLE_PERFORMANCE_PARADOXON 1
#endif

/*
 * List of libcmd builtins which do not require a |Shell_t| context.
 * This list was automatically generated from <ast/cmdext.h>
 */
static const
bfastpathrec fastpath_builtins[] =
{
	/* This list must be alphabetically sorted for |strcmp()| usage */
	{ "basename",	b_basename	},
	{ "cat",	b_cat		},
	{ "chmod",	b_chmod		},
#ifdef ENABLE_PERFORMANCE_PARADOXON
	{ "cksum",	b_cksum		},
#endif /* ENABLE_PERFORMANCE_PARADOXON */
	{ "cmp",	b_cmp		},
	{ "comm",	b_comm		},
	{ "cp",		b_cp		},
	{ "cut",	b_cut		},
	{ "date",	b_date		},
	{ "dirname",	b_dirname	},
	{ "egrep",	b_egrep		},
	{ "expr",	b_expr		},
	{ "fds",	b_fds		},
	{ "fgrep",	b_fgrep		},
	{ "fmt",	b_fmt		},
	{ "fold",	b_fold		},
	{ "getconf",	b_getconf	},
	{ "grep",	b_grep		},
	{ "head",	b_head		},
	{ "id",		b_id		},
	{ "join",	b_join		},
	{ "ln",		b_ln		},
	{ "logname",	b_logname	},
	{ "md5sum",	b_md5sum	},
	{ "mkdir",	b_mkdir		},
	{ "mkfifo",	b_mkfifo	},
	{ "mktemp",	b_mktemp	},
	{ "mv",		b_mv		},
	{ "paste",	b_paste		},
	{ "pathchk",	b_pathchk	},
	{ "pids",	b_pids		},
	{ "readlink",	b_readlink	},
#ifdef ENABLE_PERFORMANCE_PARADOXON
	{ "rev",	b_rev		},
#endif /* ENABLE_PERFORMANCE_PARADOXON */
	{ "rm",		b_rm		},
	{ "rmdir",	b_rmdir		},
	{ "stty",	b_stty		},
#ifdef ENABLE_PERFORMANCE_PARADOXON
	{ "sum",	b_sum		},
#endif /* ENABLE_PERFORMANCE_PARADOXON */
	{ "sync",	b_sync		},
	{ "tail",	b_tail		},
	{ "tee",	b_tee		},
	{ "tty",	b_tty		},
	{ "uname",	b_uname		},
	{ "uniq",	b_uniq		},
	{ "wc",		b_wc		},
	{ "xgrep",	b_xgrep		},
	{ NULL,		(int (*)(int, char **, void *))NULL }
};

static inline
const bfastpathrec *
find_bfastpathrec(const char *name)
{
	unsigned int i;
	signed int cmpres;
	for (i = 0; fastpath_builtins[i].name != NULL; i++) {
		cmpres = strcmp(fastpath_builtins[i].name, name);
		if (cmpres == 0)
			return (&fastpath_builtins[i]);
		else if (cmpres > 0)
			return (NULL);

	}
	return (NULL);
}

static inline
int
fastpath_builtin_main(const bfastpathrec *brec, int argc, char *argv[])
{
	setlocale(LC_ALL, ""); /* calls |_ast_setlocale()| */

	return ((*brec->func)(argc, argv, NULL));
}


/* Builtin script, original derived from alias.sh */
static const char *script = "\n"
/* Get name of builtin */
"typeset cmd=\"${0##*/}\"\n"
/*
 * If the requested command is not an alias load it explicitly
 * to make sure it is not bound to a path (those built-ins which
 * are mapped via shell aliases point to commands which are
 * "special shell built-ins" which cannot be bound to a specific
 * PATH element) - otherwise we may execute the wrong command
 * if an executable with the same name sits in a PATH element
 * before /usr/bin (e.g. /usr/xpg4/bin/ls would be executed
 * before /usr/bin/ls if the path was something like
 * PATH=/usr/xpg4/bin:/usr/bin).
 */
"if [[ \"${cmd}\" != ~(Elr)(alias|unalias|command) ]] && "
	"! alias \"${cmd}\" >/dev/null 2>&1 ; then\n"
	"PATH='' builtin \"${cmd}\"\n"
"fi\n"
/* command is a keyword and needs to be handled separately */
"if [[ \"${cmd}\" == \"command\" ]] ; then\n"
	"command \"$@\"\n"
"else\n"
#ifdef WORKAROUND_FOR_ALIAS_CRASH
/*
 * Work around a crash in /usr/bin/alias when invalid options are
 * passed (e.g. $ /usr/bin/alias -c #). The shell code will call
 * an error handler which does a |longjmp()| but somehow the code
 * failed to do the |setjmp()| before this point.
 * Putting the "alias" command in a subshell avoids the crash.
 * Real cause of the issue is under investigation and a fix be
 * delivered with the next ast-ksh update.
 */
	"( \"${cmd}\" \"$@\" )\n"
#else
	"\"${cmd}\" \"$@\"\n"
#endif /* WORKAROUND_FOR_ALIAS_CRASH */
"fi\n"
"exitval=$?";


static inline
int
script_builtin_main(int argc, char *argv[])
{
	int i;
	Shell_t *shp;
	Namval_t *np;
	int exitval;

	/*
	 * Create copy of |argv| array shifted by one position to
	 * emulate $ /usr/bin/sh <scriptname> <args1> <arg2> ... #.
	 * First position is set to "/usr/bin/sh" since other
	 * values may trigger special shell modes (e.g. *rsh* will
	 * trigger "restricted" shell mode etc.).
	 */
	char *xargv[argc+2];
	xargv[0] = "/usr/bin/sh";
	xargv[1] = "scriptname";
	for (i = 0; i < argc; i++) {
		xargv[i+1] = argv[i];
	}
	xargv[i+1] = NULL;

	shp = sh_init(argc+1, xargv, 0);
	if (!shp)
		error(ERROR_exit(1), "shell initialisation failed.");
	(void) sh_trap(script, 0);

	np = nv_open("exitval", shp->var_tree, 0);
	if (!np)
		error(ERROR_exit(1), "variable %s not found.", "exitval");
	exitval = (int)nv_getnum(np);
	nv_close(np);

	return (exitval);
}

int
main(int argc, char *argv[])
{
	const char *progname;
	const bfastpathrec *brec;
	char execnamebuff[PATH_MAX+1];

	/* Get program name */
	if (pathprog(argv[0], execnamebuff, sizeof (execnamebuff)) <= 0)
		error(ERROR_exit(1), "could not determinate exec name.");

	progname = (const char *)strrchr(execnamebuff, '/');
	if (progname != NULL) {
		progname++;
	}
	else
	{
		progname = execnamebuff;
	}

	/* Execute command... */
	if (brec = find_bfastpathrec(progname)) {
		/* ... either via a fast path (calling the code directly) ... */
		return (fastpath_builtin_main(brec, argc, argv));
	}
	else
	{
		/* ... or from within a full shell. */
		return (script_builtin_main(argc, argv));
	}
}
